test/extensions/transport_sockets/tls/ssl_socket_test.cc

```diff
@@ -2540,6 +2540,178 @@ TEST_P(SslSocketTest, HalfClose) {
+                                                   server_stats_store, std::vector<std::string>{});
+
+  auto socket = std::make_shared<Network::TcpListenSocket>(
+      Network::Test::getCanonicalLoopbackAddress(GetParam()), nullptr, true);
+  Network::MockTcpListenerCallbacks listener_callbacks;
+  Network::MockConnectionHandler connection_handler;
+  Network::ListenerPtr listener =
+      dispatcher_->createListener(socket, listener_callbacks, true, ENVOY_TCP_BACKLOG_SIZE);
+  std::shared_ptr<Network::MockReadFilter> server_read_filter(new Network::MockReadFilter());
+  std::shared_ptr<Network::MockReadFilter> client_read_filter(new Network::MockReadFilter());
+
+  const std::string client_ctx_yaml = R"EOF(
+    common_tls_context:
+  )EOF";
+
+  envoy::extensions::transport_sockets::tls::v3::UpstreamTlsContext tls_context;
+  TestUtility::loadFromYaml(TestEnvironment::substitute(client_ctx_yaml), tls_context);
+  auto client_cfg = std::make_unique<ClientContextConfigImpl>(tls_context, factory_context_);
+  Stats::TestUtil::TestStore client_stats_store;
+  ClientSslSocketFactory client_ssl_socket_factory(std::move(client_cfg), manager,
+                                                   client_stats_store);
+  Network::ClientConnectionPtr client_connection = dispatcher_->createClientConnection(
+      socket->localAddress(), Network::Address::InstanceConstSharedPtr(),
+      client_ssl_socket_factory.createTransportSocket(nullptr), nullptr);
+  Network::MockConnectionCallbacks client_connection_callbacks;
+  client_connection->enableHalfClose(true);
+  client_connection->addReadFilter(client_read_filter);
+  client_connection->addConnectionCallbacks(client_connection_callbacks);
+  client_connection->connect();
+
+  Network::ConnectionPtr server_connection;
+  Network::MockConnectionCallbacks server_connection_callbacks;
+  EXPECT_CALL(listener_callbacks, onAccept_(_))
+      .WillOnce(Invoke([&](Network::ConnectionSocketPtr& socket) -> void {
+        server_connection = dispatcher_->createServerConnection(
+            std::move(socket), server_ssl_socket_factory.createTransportSocket(nullptr),
+            stream_info_);
+        server_connection->enableHalfClose(true);
+        server_connection->addReadFilter(server_read_filter);
+        server_connection->addConnectionCallbacks(server_connection_callbacks);
+      }));
+  EXPECT_CALL(server_connection_callbacks, onEvent(Network::ConnectionEvent::Connected))
+      .WillOnce(Invoke([&](Network::ConnectionEvent) -> void {
+        Buffer::OwnedImpl data("hello");
+        server_connection->write(data, false);
+        // Close without sending close_notify alert.
+        const SslHandshakerImpl* ssl_socket =
+            dynamic_cast<const SslHandshakerImpl*>(server_connection->ssl().get());
+        SSL_set_quiet_shutdown(ssl_socket->ssl(), 1);
+        server_connection->close(Network::ConnectionCloseType::NoFlush);
```

Comment:
`FlushWrite` to ensure the "hello" gets sent.

Reply:
I don't think flush works while handshake is in progress.  See https://github.com/envoyproxy/envoy/issues/13890#issuecomment-721417276

Can you add some assertions regarding ssl_socket()->state() ?

Reply:
This is done after receiving `Connected` event, i.e. after the handshake, but sure, assert added.

Regarding `FlushWrite`, it appears to be broken #13890. Let me know how do you want to proceed.

Reply:
Roughly, the semantics of `close(NoFlush)` is that it will try to flush out whatever can be immediately written, then close. In this case, all egress buffers are empty, so sending this small message should always fit. So It's probably ok as-is, even though logically, we want to guarantee that the message is sent before closing, so `FlushWrite` fits the use-case better.

On the other hand, this test requires that the socket gets `shutdown(WR)` or `close(fd)` called to make sure the other side sees a TCP FIN. I don't know what BoringSSL will do exactly after `SSL_shutdown()` when quiet shutdown is set, but I am sure that completely closing the socket (as this code currently does) will make sure it is closed.

So, I think we just leave this test as-is. Except it'll have less-noisy output if you add:

`EXPECT_CALL(*server_read_filter, onNewConnection());`

Reply:
Oh, interesting... I don't see that event fired in those tests, and if I add it, then it fails locally:
```
test/extensions/transport_sockets/tls/ssl_socket_test.cc:2692: Failure
Actual function call count doesn't match EXPECT_CALL(*server_read_filter, onNewConnection())...
         Expected: to be called once
           Actual: never called - unsatisfied and active
Stack trace:
  0x45733b9: testing::internal::GoogleTestFailureReporter::ReportFailure()
  0x1c7d4e8: testing::internal::Expect()
  0x457859c: testing::internal::UntypedFunctionMockerBase::VerifyAndClearExpectationsLocked()
  0x2593480: testing::internal::FunctionMocker<>::~FunctionMocker()
  0x257d627: Envoy::Network::MockReadFilter::~MockReadFilter()
  0x257d66c: Envoy::Network::MockReadFilter::~MockReadFilter()
  0x1d289bf: std::__1::default_delete<>::operator()()
  0x1d28799: std::__1::__shared_ptr_pointer<>::__on_zero_shared()
  0x46e167e: std::__1::__shared_weak_count::__release_shared()
  0x1b837f9: Envoy::Extensions::TransportSockets::Tls::SslSocketTest_ShutdownWithoutCloseNotify_Test::TestBody()
  0x45c0744: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x45af4fb: testing::internal::HandleExceptionsInMethodIfSupported<>() 
  0x459afd3: testing::Test::Run()
  0x459b9a7: testing::TestInfo::Run()
... Google Test internal frames ...
```

Reply:
Weird; I ran into that locally. Given that it is passing CI, let's merge it as-is

Reply:
Close(FlushWrite) does not work because of this code:

https://github.com/envoyproxy/envoy/blob/7d6e7a4e559bdf0346687f7f404412e2412ea6fb/source/extensions/transport_sockets/tls/ssl_handshaker.cc#L231

The onEvent mock function runs from the handshake_callbacks_->onSuccess(ssl()); call.  The call to close results in doHandshake returning a close PostIoAction that triggers a "remote close" event on the server connection.  We do get a onData("", true) call on the client connection later, but the "hello" string is lost.

Use of mocks is once again a big source of confusion.

Reply:
Thanks for investigating! I fixed the test to make it more representative of the production code: #13912.